### ISSUE TITLE
pkg/sql/logictest/tests/local: `TestLogic_record` fails under `race`

### ISSUE THREAD

**Issue by rickystewart** [cockroachdb/cockroach#118691]
[EngFlow link](https://tanzanite.cluster.engflow.com/invocations/default/928a3274-593b-4ac1-a27e-0975c02c7809?testReportRun=1&testReportAttempt=0#targets-Ly9wa2cvc3FsL2xvZ2ljdGVzdC90ZXN0cy9sb2NhbDpsb2NhbF90ZXN0)

```
[09:52:18] setting distsql_workmem='58834B';
==================
WARNING: DATA RACE
Read at 0x00c001e4a0a0 by goroutine 33175:
  github.com/cockroachdb/cockroach/pkg/sql/types.(*T).Size()
      github.com/cockroachdb/cockroach/pkg/sql/types/pkg/sql/types/types.go:2157 +0x45
  github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.(*ColumnDescriptor).Size()
      github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb/bazel-out/k8-fastbuild/bin/pkg/sql/catalog/descpb/descpb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb/structured.pb.go:10250 +0x117
  github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb.(*TableDescriptor).Size()
      github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb/bazel-out/k8-fastbuild/bin/pkg/sql/catalog/descpb/descpb_go_proto_/github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb/structured.pb.go:10675 +0x3e8
  github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc.(*wrapper).ByteSize()
      github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc/table_desc.go:135 +0x31
  github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc.(*immutable).ByteSize()
```

Jira issue: CRDB-35844

**Comment by DrewKimball** [comment]
I'd expect this to be fixed by #118585, which merged to master yesterday. Is there a way to see which commit this test failed on?

**Comment by DrewKimball** [comment]
Nevermind, I found it. It looks like #118585 wasn't enough because it doesn't handle the recursive type references in tuple and array types.
